contract/src/stage1/delete.rs

```diff
@@ -1,40 +1,223 @@
+            let promise = env::promise_batch_create(&env::current_account_id());
+
+            env::promise_batch_action_delete_key(
+                promise, 
+            &key, 
+            );
+
+            env::promise_return(promise);
+        }
+    }
+
+    /*
+        Refund NFTs or FTs for a drop. User can optionally pass in a number of assets to
+        refund. If not, it will try to refund all assets.
+    */
+    pub fn refund_assets(&mut self, 
+        drop_id: DropId,
+        assets_to_refund: Option<u64>
+    ) {
+        // get the drop object
+        let mut drop = self.drop_for_id.get(&drop_id).expect("No drop found");
+        let funder_id = drop.funder_id.clone();
+        require!(funder_id == env::predecessor_account_id(), "only drop funder can delete keys");
+
+        // Get the number of keys registered for the drop.
+        let keys_registered = drop.keys_registered;
+        require!(keys_registered > 0, "no keys left to unregister");
         
-//         // If there's a Nonce, remove all occurrences of the nonce
-//         if let Some(nonce) = account_data.cb_id {
-//             self.ft.remove(&nonce);
-//             self.nft.remove(&nonce);
-//             self.fc.remove(&nonce);
-//         }
+        // Get the keys to refund. If not specified, this is the number of keys currently registered.
+        let num_to_refund = assets_to_refund.unwrap_or(keys_registered);
+        require!(num_to_refund <= keys_registered, "can only refund less than or equal to the amount of keys registered");
+
+        // Decrement the drop's keys registered temporarily. If the transfer is unsuccessful, revert in callback. 
+        drop.keys_registered -= num_to_refund;
+        self.drop_for_id.insert(&drop_id, &drop);
         
-//         // Delete the key
-//         Promise::new(env::current_account_id())
-//             .delete_key(public_key);
-
-//         // balance.0 + ACCESS_KEY_ALLOWANCE + required_storage + ACCESS_KEY_STORAGE + if fc_data.is_some() {fc_data.clone().unwrap().deposit.0} else {0}
-//         // ACCESS_KEY_STORAGE + required_storage.0 + ACCESS_KEY_ALLOWANCE + balance.0 + min.0
-//     }
-// }
\ No newline at end of file
+        // Only NFT or FT assets can be refunded
+        if let Some(data) = drop.nft_data {
+            /*
+                NFTs need to be batched together. Loop through and transfer all NFTs.
+                Keys registered will be decremented and the token IDs will be removed
+                in the callback if everything is successful. If anything fails, the 
+                keys registered will be added back in the callback for the drop.
+            */ 
+            let nft_batch_index = env::promise_batch_create(&data.nft_contract);
+            let token_ids: Vec<String> = data.token_ids.unwrap().iter().take(num_to_refund.try_into().unwrap()).collect();
+            require!(token_ids.len() as u64 == num_to_refund, "not enough token IDs");
```

Comment:
Where is the upper limit? What's stopping the user from passing in 1000 or None? The call will fail and keys_registered will revert, but where's the guidance for the client? Do we have a limit in mind for NFTs?

Reply:
Talked about this in TG - it's fine to level it for now. If they pass in too many, the GAS calculations will fail and none of the promises will fire.

Reply:
Agree